### PR TITLE
Support shuffling the total dataset.

### PR DESCRIPTION
**Pull request by workingloong**

### CODE REVIEW

File: elasticai_api/proto/elasticai_api.proto [hunk 1]
```diff
@@ -24,6 +24,7 @@ message Shard {
   // Starting and ending (non-inclusive) record number.
   int64 start = 2;
   int64 end = 3;
+  repeated int64 indices = 4;
```
Comment (brightcoder01) [inline elasticai_api/proto/elasticai_api.proto:27]: I think we can make the shard size small and then shuffle the shard list instead of shuffling each id and returning the id list.

Comment (workingloong) [reply elasticai_api/proto/elasticai_api.proto:27]: We usually store the images in the folder with the class name. For example:
```
|--mnist_png
    |--0
        |--001.png
        |--002.png
        |...
    |--1
       |--001.png
       |--002.png
```

Then, we can use `torchvision.datasets.ImageFolder("./mnist_png")` to load paths and classes of the images. 
image

If we create shards according to the `dataset.imgs`, the shard may only contains samples of a class. So we must shuffle all indices to create shards.

File: elasticai_api/common/data_shard_service.py [hunk 2]
```diff
@@ -45,12 +44,14 @@ def __init__(
...
         num_epochs=None,
         dataset_size=None,
+        shuffle=False,
         task_type=elasticai_api_pb2.TRAINING,
     ):
         self._mc = master_client
         self._batch_size = batch_size
         self._num_epochs = num_epochs
         self._dataset_size = dataset_size
+        self._shuffle = shuffle
```
Comment (brightcoder01) [inline elasticai_api/common/data_shard_service.py:54]: I think we can name it element_wise_shuffle or per_sample_shuffle.
Because we will shuffle the shard list for training by default.

File: elasticdl/python/master/task_manager.py [hunk 3]
```diff
@@ -293,6 +308,11 @@ def create_tasks(self, task_type, model_version=-1):
                     start_idx_this_task + self._records_per_task,
                     max_idx_this_shard,
                 )
+                task_record_indices = task_record_indices = (
```
Comment (brightcoder01) [inline elasticdl/python/master/task_manager.py]: `task_record_indices = task_record_indices = xxx`.
Is it a typo?

Comment (workingloong) [reply elasticdl/python/master/task_manager.py]: Yes, fix it!

File: elasticai_api/common/data_shard_service.py (hunk 2 above)
Comment (workingloong) [reply elasticai_api/common/data_shard_service.py:54]: `shuffle` is a common word in PyTorch and TensorFlow. Here, I want to keep the same as these frameworks and not introduce a new word to users. I think we should use the argument `shuffle_shards` to shuffle the shard list for training.

Comment (brightcoder01) [reply elasticai_api/common/data_shard_service.py:54]: `shuffle_shards` is a good idea.

Comment (workingloong) [reply elasticai_api/common/data_shard_service.py:54]: Yeah, and we can add the argument `shuffle_shards` in the next PR.

**Review by brightcoder01 — APPROVED** [review]
